### DISCUSSION TITLE
Use of inline r code in code-chunk options fig-cap and tbl-cap

### DISCUSSION THREAD

**Discussion by larnsce** [quarto-dev/quarto-cli#6193] [Q&A] [answered]
### Description

I would like to use dynamic figure and table captions in quarto documents, e.g.

```
---
format: html
params
  year: 2022
---
```

In `@fig-requests-year` the ...

```
```{r}
#| label: fig-requests-year
#| fig-cap: "Number of requests per year in `r params$year`."

fig_requests
```

**Comment by mcanouil** [comment]
You have to use Quarto syntax as described in <https://quarto.org/docs/computations/r.html#chunk-options>.

```r
#| fig-cap: !expr 'paste("Air", "Quality")'
```

I am changing this discussion to Q&A.

**Reply by larnsce** [reply]
Thank you for your quick and helpful feedback.

**Reply by jkylearmstrong** [reply]
Hey @mcanouil, I am very confused by this note: 

> the !expr syntax is an example of a YAML “tag” literal, and it can be unintuitive. !expr needs to be followed by a single YAML “flow scalar”: see the [YAML spec](https://yaml.org/spec/1.2.2/#73-flow-scalar-styles) for details on how double-quoted, single-quoted, and unquoted strings work.

when I am making a `fig-cap` or `tbl-cap` within a child document, I might want to reference a corresponding table or figure in addition to using an r code value?


Would the corresponding code look like this:
```
#| label: tbl-{{var}}
#| fig-cap: !expr 'paste("Corresponding to ", @fig-{{var}}, " we find the mean is, ", mean_value , ".)'
```  

A few more examples would be helpful in the documentation.

Thanks!

**Reply by mcanouil** [reply, marked as answer]
This thread is very old.
As with recent version of Quarto, I would highly recommend to avoid `!expr` which is knitr specific and can lead to very weird syntax.

Use the [cross-reference div syntax](https://quarto.org/docs/authoring/cross-references-divs.html) and [inline code](https://quarto.org/docs/computations/inline-code.html).

````qmd
::: {#<prefix>-<whatever-name>}

```{language}
Code
```

The caption with `{language} code`.
::: 
````


`@fig-{{var}}` is not an R expression in your case.
What you provide to `!expr` still needs to be valid R code.

**Comment by baptiste** [comment]
You can also use the older knitr syntax, which is less constrained by Yaml limitations

````
---
title: "Untitled"
format: html
params:
  year: 2022
---

In @fig-requests-year the ...

```{r, fig.cap=glue::glue("Number of requests per year in {params$year}.")}
#| label: fig-requests-year

plot(1,1)
```


````

**Reply by mcanouil** [reply]
You should not mixed YAML and inline code cell syntax, even in `rmarkdown` it is not recommended.

Also Quarto recommends YAML syntax which is the syntax fully supported.

**Comment by jkylearmstrong** [comment]
Thanks @mcanouil I'll try it out, I understand the post is old but it's also where google directed me! So hopefully, it will help others in the future.

**Reply by mcanouil** [reply]
It's been a very long time since I relied on Google indexing to find something.
GitHub has a pretty good (in my opinion) search engine and here, on Quarto CLI repository, we try to label everything to help finding useful tips, workaround, etc., so I suggest to use it for Quarto related matters.
I strongly think the best resources for Quarto are:
- GitHub Discussions
- https://quarto.org
- GitHub Issues
